### PR TITLE
Project file documentation

### PR DESCRIPTION
**Pull request by chleh**
Cf. Issue #976.

I played around a little bit with the project file documentation. I think it is best to do it entirely within doxygen, since this way everybody can build the docs locally.

Currently this PR:
* adds a doxygen alias that makes a list of all used parameters
* adds a hierarchy in Documentation/ProjectFile that resembles the tree structure of the xml file; there the actual documentation sits
* ~~uses ProjectData.cpp as an example~~

* [x] Please tell me, if that approach fits our needs, and about possible improvements (automatic check scripts will follow).

TODO:
* [x] Write some general remarks how to read the docs and how to document parameters.
* [x] Clean up scripts.

Future Work:
* Fill the documentation files with meaningful content.
* Provide a script s.t. Jenkins can test for correct documentation.
* Detect polymorphic types in end-to-end test input files. Provide a list of untested configuration options in the quality assurance page.

### CODE REVIEW

**Comment by chleh** [comment]
The result currently looks like that:

param-docs

**Comment by chleh** [comment]
The idea to to later on have the hierarchy automatically created—so one does not need to fiddle around with long identifier names—and to have only plain markdown files in the documentation tree that will be automatically augmented, e.g., with links to benchmarks.

**Comment by bilke** [comment]
I like it but I am also curious how you automate the hierarchy creation..

**Comment by TomFischer** [comment]
Nice!

**Comment by wenqing** [comment]
:+1:

**Comment by chleh** [comment]
@bilke: challenge accepted ^^

**Comment by chleh** [comment]
Current screenshots.

Augmented documentation. (Needs python and bash. Note: The basic documentation, i.e., the tree on the left and the handwritten content of the doc pages only need cmake and doxygen.)
input-doc-1

Documentation can handle XML tags, XML attributes and "special cases" like GWFlow whose parameters are not written verbatim in the `getConfParam()` calls:
input-doc-2

**Comment by chleh** [comment]
@rinkk The scripts I wrote provide pretty much information about what a valid project file looks like. I think, they can be extended without much effort to automatically generate an XSD file for our prj files, and furthermore to always keep that XSD file always in sync with our code. However, there might be some cornercases with the XSD that require some extra work.

**Comment by endJunction** [comment]
Are those `.md` files generated automatically?
I understand that the content of those `.md` files will be the actual documentation.
What happens if the parameter is removed from the code?

@bilke Could you please have a glance on the cmake scripts? Maybe you see some obvious artifacts there...

@ all I think we shall merge this and see how it works out regarding the documenting efforts; and the second opinion will come from the users.

:fast_forward:

File: scripts/cmake/DocumentationSetup.cmake [hunk 1]
```diff
@@ -38,4 +38,36 @@ if(DOXYGEN_FOUND)
 
     configure_file(Documentation/Doxyfile.in ${PROJECT_BINARY_DIR}/Doxyfile)
 
+    if (NOT BASH_TOOL_PATH STREQUAL "BASH_TOOL_PATH-NOTFOUND"
+            AND NOT PYTHON_EXECUTABLE STREQUAL "PYTHON_EXECUTABLE-NOTFOUND")
```
Comment (bilke) [inline scripts/cmake/DocumentationSetup.cmake]: The string comparison is not necessary. Simply check like this:

```cmake
if (NOT BASH_TOOL_PATH AND NOT PYTHON_EXECUTABLE)
```

From the docs:

> False if the constant is 0, ... or ends in the suffix -NOTFOUND.

File: scripts/cmake/DocumentationSetup.cmake [hunk 2]
```diff
@@ -38,4 +38,36 @@ if(DOXYGEN_FOUND)
...
+        set(doc_use_external_tools TRUE)
+    else()
+        set(doc_use_external_tools FALSE)
```
Comment (bilke) [inline scripts/cmake/DocumentationSetup.cmake]: I guess the setting of the variable should be the other way around? 

Also if this var is just used once I would put the tools check into the if-command below and remove the variable.

File: scripts/doc/check-project-params.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+for inline in sys.stdin:
+    inline = inline.strip().split("@@@")
+    status = inline[0]
+
+    if status == "OK":
+        tag_path_comment = inline[3]
+        tag_name_comment = tag_path_comment.split(".")[-1]
+
+        dirs = tag_path_comment.split(".")[:-1]
+        p = os.path.join(docauxdir, *dirs, )
```
Comment (bilke) [inline scripts/doc/check-project-params.py]: Is the `,` after dirs correct?

**Comment by chleh** [comment]
> Are those .md files generated automatically?

No, but there is the script `create-docu-file-stubs.sh`, which can be used to create missing md files. This script should cover most of the cases (not all) and reduces the number of files and directories that have to be created manually.

> I understand that the content of those .md files will be the actual documentation.

Excatly.

> What happens if the parameter is removed from the code?

There is not yet a check if an md file is obsolete. I'll add one. Then it will be indicated on the QA page.

File: scripts/cmake/DocumentationSetup.cmake (hunk 1 above)
Comment (chleh) [inline scripts/cmake/DocumentationSetup.cmake]: Thanks for pointing that out.

**Comment by bilke** [comment]
Now I get the following error on Mac OS, Python 2.7.11:

```
[100%] Generating project file documentation quality assurance pages.
readlink: illegal option -- f
usage: readlink [-n] [file ...]
grep: -: No such file or directory
expr: syntax error
  File "/Users/bilke/code/ogs6/ogs/scripts/doc/check-project-params.py", line 117
    + "| [&rarr; ufz/ogs/master]({6}/{0}#L{1})").format(*u, github_src_url))
SyntaxError: only named arguments may follow *expression
  File "/Users/bilke/code/ogs6/ogs/scripts/doc/append-xml-tags.py", line 134
    tagpath_expanded = ".".join((tagpathhead, *tagpathparts[1:])).lstrip(".")
                                              ^
SyntaxError: invalid syntax
```

**Comment by chleh** [comment]
I hope my last commits fix the python2 and the shell issues.
Btw. the python and bash detected by cmake are _not_ used. Is that a problem?

**Comment by endJunction** [comment]
Jenkins, test this please!

**Comment by endJunction** [comment]
@chleh Any chances to get this done before Thursday? I'll contribute the documentation (of two-three tags).

**Comment by endJunction** [comment]
Maybe Jenkins is in the "not finished" state because it can not merge with current master. A rebase would then trigger the tests, isn't it?

**Comment by chleh** [comment]
I just rebased.

**Comment by endJunction** [comment]
@chleh Some plattforms don't have `realpath` https://svn.ufz.de:8443/view/OGS-6%20Monitor/job/OGS-6/job/Docs/617/console
Maybe we need to change that to `readlink -f -- "$@"` ?

**Comment by chleh** [comment]
@endJunction, I changed that to realpath, because @bilke reported the following error message on Mac: `readlink: illegal option -- f`

**Comment by chleh** [comment]
Python 2.6.6 complains that it does not know argparse. Maybe add the line `module load python/2/7.9-3` to the Jenkins job? Of course, one could also remove the argparse dependency.

**Comment by chleh** [comment]
I probably fixed the issues in 9c6eec6f32a6dd47aedf5aa5e8cc0a9e56f8de71.

**Pull request merged**